<img src=".\images\Logo.png" width=150 align="left" /> <img src=".\images\Logo2.jpg" width=450 align="right" />


# <center><font color= #1e8449 > <b>CUxx_xxxxx</font></center>

<font color='steelblue'><b>Citizenlab Data Science Methodology > II - Data Processing Domain </font>
***
> # <font color='steelblue'> <b>05.- Data Collection</font>

<font color='steelblue'>Data Collection is the process to obtain and generate (if required) necessary data to model the problem.</font>

### 16. Unión de datos semanales por zonas de vacunación y otras

* Consolidación de todas las variables por semana del caso de uso
* Vacunación, meteo, contaminación







<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Settings" data-toc-modified-id="Settings-1"><font color="green">Settings</font></a></span></li><li><span><a href="#Data-Load" data-toc-modified-id="Data-Load-2"><font color="green">Data Load</font></a></span></li><li><span><a href="#ETL-Processes" data-toc-modified-id="ETL-Processes-3">ETL Processes</a></span><ul class="toc-item"><li><span><a href="#Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files" data-toc-modified-id="Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files-3.1">Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files</a></span></li></ul></li><li><span><a href="#Synthetic-Data-Generation" data-toc-modified-id="Synthetic-Data-Generation-4">Synthetic Data Generation</a></span></li><li><span><a href="#Fake-Data-Generation" data-toc-modified-id="Fake-Data-Generation-5">Fake Data Generation</a></span></li><li><span><a href="#Open-Data" data-toc-modified-id="Open-Data-6">Open Data</a></span></li><li><span><a href="#Data-Save" data-toc-modified-id="Data-Save-7"><font color="green">Data Save</font></a></span></li><li><span><a href="#Main-Conclusions" data-toc-modified-id="Main-Conclusions-8"><font color=" #2874a6 ">Main Conclusions</font></a></span></li><li><span><a href="#Main-Actions" data-toc-modified-id="Main-Actions-9"><font color=" #2874a6 ">Main Actions</font></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Acciones-done" data-toc-modified-id="Acciones-done-9.0.1"><font color=" #2874a6 ">Acciones done</font></a></span></li><li><span><a href="#Acctions-to-perform" data-toc-modified-id="Acctions-to-perform-9.0.2"><font color=" #2874a6 ">Acctions to perform</font></a></span></li></ul></li></ul></li></ul></div>

## <font color='green'>Settings</font>

### Packages to use


- {tcltk} para selección interactiva de archivos locales
- {readxl} para leer archivos de Excel
- {readr} para leer y escribir archivos csv
- {dplyr} para explorar datos



In [41]:
library(readxl)
library(readr)
library(dplyr)




### Paths

In [6]:
iPath <- "Data/Input/"
oPath <- "Data/Output/"

## <font color='green'>Data Load</font>

If there are more than one input file, make as many sections as files to import.

Instrucciones
- Los ficheros de entrada del proceso están siempre en Data/Input/.   
- Si hay más de un fichero de entrada, se crean tantos objetos iFile_xx y file_data_xx como ficheros de entrada (xx número correlativo con dos dígitos, rellenar con ceros a la izquierda)

1. Datos de vacunación y capacidad

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [7]:
# file_data_01 <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [8]:
iFile_01 <- "Output.xlsx"
file_data_01 <- paste0(iPath, iFile_01)

if(file.exists(file_data_01)){
    cat("Se leerán datos del archivo: ", file_data_01)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leer<U+00E1>n datos del archivo:  Data/Input/Output.xlsx

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [9]:
data_01 <- read_excel(file_data_01)


Estructura de  los datos:

In [10]:
data_01 |> glimpse()

Rows: 20,868
Columns: 22
$ ID                   <chr> "1", "2", "3", "4", "5", "6", "7", "8", "9", "10"~
$ CODBDT               <chr> "686213", "686213", "686213", "686213", "686213",~
$ GEOCODIGO            <chr> "001", "001", "001", "001", "001", "001", "001", ~
$ DESBDT               <chr> "Abrantes", "Abrantes", "Abrantes", "Abrantes", "~
$ semana               <dbl> 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 4~
$ ano                  <dbl> 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2~
$ n_vacunas            <dbl> 0, 0, 0, 0, 0, 328, 344, 353, 371, 341, 389, 349,~
$ n_citas              <dbl> 0, 0, 0, 0, 0, 305, 327, 328, 341, 334, 373, 332,~
$ capacidad_zona       <dbl> 7742, 7654, 7425, 7189, 7566, 7417, 7638, 7711, 7~
$ prop_riesgo          <dbl> 0.1954797, 0.1906348, 0.1930880, 0.1918070, 0.185~
$ tasa_riesgo          <dbl> 0.1950852, 0.1870784, 0.1874322, 0.2031197, 0.188~
$ poblacion_total      <dbl> 29872, 29872, 29872, 29872, 29872, 29872, 29872, ~
$ poblacion_muj

Muestra de datos:

In [11]:
data_01 |> slice_head(n = 5)

ID CODBDT GEOCODIGO DESBDT   semana ano  n_vacunas n_citas capacidad_zona
1 1  686213 001       Abrantes 36     2021 0         0       7742          
2 2  686213 001       Abrantes 37     2021 0         0       7654          
3 3  686213 001       Abrantes 38     2021 0         0       7425          
4 4  686213 001       Abrantes 39     2021 0         0       7189          
5 5  686213 001       Abrantes 40     2021 0         0       7566          
  prop_riesgo ... poblacion_mujeres poblacion_mayores poblacion_inmigrante
1 0.1954797   ... 0.5345094         0.1781619         0.2183993           
2 0.1906348   ... 0.5345094         0.1781619         0.2183993           
3 0.1930880   ... 0.5345094         0.1781619         0.2183993           
4 0.1918070   ... 0.5345094         0.1781619         0.2183993           
5 0.1851164   ... 0.5345094         0.1781619         0.2183993           
  tasa_paro tasa_mayores temperatura humedad NO2 busquedas_gripe tuits_gripe
1 0.1665607 0.1644978    NA          NA      NA  NA              0          
2 0.1665607 0.1882310    NA          NA      NA  NA              0          
3 0.1665607 0.1940977    NA          NA      NA  NA              0          
4 0.1665607 0.1781833    NA          NA      NA  NA              1          
5 0.1665607 0.1814519    NA          NA      NA  NA              0

2. Datos meteorológicos

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [12]:
# file_data_02 <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [13]:
iFile_02 <- "CU_04_05_11_meteo_zonas.csv"
file_data_02 <- paste0(iPath, iFile_02)

if(file.exists(file_data_02)){
    cat("Se leerán datos del archivo: ", file_data_02)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leer<U+00E1>n datos del archivo:  Data/Input/CU_04_05_11_meteo_zonas.csv

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [14]:
data_02 <- read_csv(file_data_02)


Rows: 21736 Columns: 7
-- Column specification --------------------------------------------------------
Delimiter: ","
chr (1): GEOCODIGO
dbl (6): tmed, prec, velmedia, presMax, ano, semana

i Use `spec()` to retrieve the full column specification for this data.
i Specify the column types or set `show_col_types = FALSE` to quiet this message.


Estructura de  los datos:

In [15]:
data_02 |> glimpse()

Rows: 21,736
Columns: 7
$ GEOCODIGO <chr> "001", "002", "003", "004", "005", "006", "007", "008", "009~
$ tmed      <dbl> 22.53414, 22.42739, 22.44443, 22.37906, 21.83238, 22.57096, ~
$ prec      <dbl> 1.2086417, 1.2461203, 1.2778878, 1.2615862, 1.2790072, 1.019~
$ velmedia  <dbl> 2.299920, 2.310469, 2.382023, 2.321377, 2.461297, 2.029950, ~
$ presMax   <dbl> 940.7107, 938.2132, 941.6089, 939.4657, 951.1882, 944.2638, ~
$ ano       <dbl> 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, ~
$ semana    <dbl> 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, ~


Muestra de datos:

In [16]:
data_02 |> slice_head(n = 5)

GEOCODIGO tmed     prec     velmedia presMax  ano  semana
1 001       22.53414 1.208642 2.299920 940.7107 2021 35    
2 002       22.42739 1.246120 2.310469 938.2132 2021 35    
3 003       22.44443 1.277888 2.382023 941.6089 2021 35    
4 004       22.37906 1.261586 2.321377 939.4657 2021 35    
5 005       21.83238 1.279007 2.461297 951.1882 2021 35

3. Datos de contaminación

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [17]:
# file_data_03 <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [18]:
iFile_03 <- "CU_04_05_15_zonas_contaminacion.csv"
file_data_03 <- paste0(iPath, iFile_03)

if(file.exists(file_data_03)){
    cat("Se leerán datos del archivo: ", file_data_03)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leer<U+00E1>n datos del archivo:  Data/Input/CU_04_05_15_zonas_contaminacion.csv

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [19]:
data_03 <- read_csv(file_data_03)


Rows: 21164 Columns: 12
-- Column specification --------------------------------------------------------
Delimiter: ","
chr  (1): GEOCODIGO
dbl (11): benzene, co, no, no2, nox, o3, pm10, pm2.5, so2, ano, semana

i Use `spec()` to retrieve the full column specification for this data.
i Specify the column types or set `show_col_types = FALSE` to quiet this message.


Estructura de  los datos:

In [20]:
data_03 |> glimpse()

Rows: 21,164
Columns: 12
$ GEOCODIGO <chr> "001", "002", "003", "004", "005", "006", "007", "008", "009~
$ benzene   <dbl> 0.2478896, 0.2351470, 0.2438505, 0.2311297, 0.2582252, 0.326~
$ co        <dbl> 0.2783157, 0.2720947, 0.2564127, 0.2704844, 0.3730037, 0.422~
$ no        <dbl> 4.277267, 4.277267, 4.277267, 4.277267, 4.277267, 4.277267, ~
$ no2       <dbl> 30.81135, 30.34314, 30.12568, 30.05632, 25.87842, 27.88230, ~
$ nox       <dbl> 38.70242, 38.26907, 37.77193, 37.77965, 32.30093, 35.32128, ~
$ o3        <dbl> 53.41865, 55.30295, 55.55106, 57.16719, 57.60870, 59.73277, ~
$ pm10      <dbl> 17.55392, 21.35499, 22.63530, 23.80635, 22.43009, 17.30619, ~
$ pm2.5     <dbl> 12.77851, 12.84574, 12.85695, 12.86689, 12.18365, 12.24123, ~
$ so2       <dbl> 6.747270, 7.411119, 6.654251, 7.439685, 2.376795, 1.334976, ~
$ ano       <dbl> 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, ~
$ semana    <dbl> 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, ~


Muestra de datos:

In [21]:
data_03 |> slice_head(n = 5)

GEOCODIGO benzene   co        no       no2      nox      o3       pm10    
1 001       0.2478896 0.2783157 4.277267 30.81135 38.70242 53.41865 17.55392
2 002       0.2351470 0.2720947 4.277267 30.34314 38.26907 55.30295 21.35499
3 003       0.2438505 0.2564127 4.277267 30.12568 37.77193 55.55106 22.63530
4 004       0.2311297 0.2704844 4.277267 30.05632 37.77965 57.16719 23.80635
5 005       0.2582252 0.3730037 4.277267 25.87842 32.30093 57.60870 22.43009
  pm2.5    so2      ano  semana
1 12.77851 6.747270 2021 35    
2 12.84574 7.411119 2021 35    
3 12.85695 6.654251 2021 35    
4 12.86689 7.439685 2021 35    
5 12.18365 2.376795 2021 35

## ETL Processes

### Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files

Se han importado en el apartado Data Load anterior:

* Datos simulados de vacunación y otros
* Datos meteorológicos
* Datos de contaminación

Incluir apartados si procede para: Extracción de datos (select, filter), Transformación de datos, (mutate, joins, ...). Si es necesario tratar datos perdidos, indicarlo también en NB 09.2

Si no aplica: Estos datos no requieren tareas de este tipo.

#### Data extract

* Seleccionar columnas de la simulación que son por semana

In [23]:
edata_01 <- data_01 |> 
    select(GEOCODIGO, DESBDT, ano, semana, n_vacunas,
        n_citas)

#### Data transform

* Unión de tablas por semana y zona

In [34]:
data <- edata_01 |>
    full_join(data_02, by = c("GEOCODIGO", "ano", "semana")) |>
    full_join(data_03, by = c("GEOCODIGO", "ano", "semana"))

In [35]:
glimpse(data)

Rows: 21,736
Columns: 19
$ GEOCODIGO <chr> "001", "001", "001", "001", "001", "001", "001", "001", "001~
$ DESBDT    <chr> "Abrantes", "Abrantes", "Abrantes", "Abrantes", "Abrantes", ~
$ ano       <dbl> 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, ~
$ semana    <dbl> 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, ~
$ n_vacunas <dbl> 0, 0, 0, 0, 0, 328, 344, 353, 371, 341, 389, 349, 402, 350, ~
$ n_citas   <dbl> 0, 0, 0, 0, 0, 305, 327, 328, 341, 334, 373, 332, 370, 336, ~
$ tmed      <dbl> 23.822231, 20.160014, 18.551058, 18.815387, 17.494475, 17.56~
$ prec      <dbl> -0.0063023484, 3.5537258042, 3.9769052178, 1.4806472324, -0.~
$ velmedia  <dbl> 3.573728, 2.494744, 3.316148, 2.384262, 1.850839, 1.527085, ~
$ presMax   <dbl> 940.9841, 940.7610, 943.1540, 944.6697, 944.2973, 943.0733, ~
$ benzene   <dbl> 0.1713567, 0.1573829, 0.1858059, 0.1486437, 0.1428037, 0.141~
$ co        <dbl> 0.1680325, 0.2138607, 0.2034376, 0.2399882, 0.2693450, 0.283~
$ no        <db

## Synthetic Data Generation

Algunos de los datos se habían generado de forma sintética

## Fake Data Generation

No aplica


## Open Data

Las zonas de salud se habían obtenido de fuentes abiertas

## <font color='green'>Data Save</font>

Este proceso, puede copiarse y repetirse en aquellas partes del notebbok que necesiten guardar datos.
Recuerde cambiar las cadenas añadida del fichero para diferenciarlas

<font color='tomato'> Identificamos los datos a guardar</font>

In [36]:
data_to_save <- data


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [37]:
caso <- "CU_04"
proceso <- '_05'
tarea <- "_16"
archivo <- ""
proper <- "_vacunacion_gripe"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [ ]:
# file_save <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out <- paste0(oPath, file_save)
# write_csv(data_to_save, path_out)

# cat('File saved as: ')
# path_out

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [38]:
file_save <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out <- paste0(oPath, file_save)
write_csv(data_to_save, path_out)

cat('File saved as: ')
path_out

File saved as: 

[1] "Data/Output/CU_04_05_16_vacunacion_gripe.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [39]:
path_in <- paste0(iPath, file_save)
file.copy(path_out, path_in, overwrite = TRUE)

[1] TRUE

## <font color=' #2874a6 '>Main Conclusions</font>

<font color='steelblue'>List and describe the general conclusions of the analysis carried out​.</font>

### Prerequisites

Para que funcione este código se necesita:

* Las rutas de archivos `Data/Input` y `Data/Output` deben existir (relativas a la ruta del _notebook_)
* El paquete tcltk instalado para seleccionar archivos interactivamente. No se necesita en producción.
* Los paquetes tcltk, readxl, readr, dplyr deben estar instalados.

### Configuration Management

This notebook has been tested with the following versions of R and packages. It cannot be assured that later versions work in the same way:
* R 4.2.2
* tcltk 4.2.2
* readxl 1.4.1
* readr 2.1.3
* dplyr 1.0.10

### Data structures

#### Objeto `data`

- Los datos de origen 
- Hay 21736 filas con la información de las siguientes variables:
	* GEOCODIGO
	* DESBDT
	* ano
	* semana
	* n_vacunas
	* n_citas
	* tmed
	* prec
	* velmedia
	* presMax
	* benzene
	* co
	* no
	* no2
	* nox
	* o3
	* pm10
	* pm2.5
	* so2


#### Observaciones generales sobre los datos

* Había más datos de vacunación que de las otras variables porque las primeras se simularon de las semanas exactas de la campaña, y el resto se obtuvieron de meses completos
* Se han mantenido todas las filas en este conjunto de datos por si son útiles para los modelos
* Los datos son únicos para cada zona, año y semana

### Consideraciones para despliegue en piloto

* Ninguna

### Consideraciones para despliegue en producción

* Se deben crear los procesos ETL en producción necesarios para que los datos de entrada estén actualizados

## <font color=' #2874a6 '>Main Actions</font>

#### <font color=' #2874a6 '>Acciones done</font>

<font color='steelblue'>Indicate the actions that have been carried out in this process</font>

- Se han unido las tablas con datos por semana y zona

#### <font color=' #2874a6 '>Acctions to perform</font>

<font color='steelblue'>Indicate the actions that must be carried out in subsequent processes</font>

- Se deben unir a las tablas con otra granularidad, repitiendo datos, para tener un csv que incluya todo

## <font color='RED'>CODE TO DEPLOY (PILOT)</font>

A continuación se incluirá el código que deba ser llevado a despliegue para producción, dado que se entiende efectúa operaciones necesarias sobre los datos en la ejecución del prototipo

<font color=' red '><b>Description</b></font>

- No hay nada que desplegar en el piloto, ya que estos datos son estáticos o en todo caso cambian con muy poca frecuencia, altamente improbable durante el proyecto.


<font color=' red '><b>CODE</b></font>

In [ ]:
# incluir código